File: src/knapsack_branch_bound.cpp
/**
 * Branch and Bound Solver for 0-1 Knapsack Problem
 *
 * Replaces the O(2^n) brute force enumeration with a DFS Branch and Bound
 * search guided by a linear (LP) relaxation upper bound.
 *
 * Pipeline:
 *     1. Items are pre-sorted by value/weight ratio descending.
 *     2. DFS tries "include" before "exclude" so a strong incumbent appears early.
 *     3. Bound = current_value + fractional knapsack over remaining items.
 *        If bound <= best_value, the subtree is pruned.
 *     4. Infeasible (weight > capacity) branches are pruned immediately.
 *
 * Worst case is still O(2^n), but the bound typically collapses the search
 * by orders of magnitude on non-pathological instances.
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <chrono>
#include <iomanip>
#include <algorithm>
#include <numeric>

using namespace std;

struct KnapsackInstance {
    int n_items;
    int capacity;
    vector<int> weights;
    vector<int> values;
};

struct Solution {
    int max_value;
    int total_weight;
    vector<int> selected_items;
    double time_seconds;
};

KnapsackInstance read_knapsack_instance(const string& filepath) {
    KnapsackInstance instance;
    ifstream file(filepath);

    if (!file.is_open()) {
        cerr << "Error: Could not open file " << filepath << endl;
        exit(1);
    }

    string line;
    vector<string> lines;
    while (getline(file, line)) {
        line.erase(0, line.find_first_not_of(" \t\r\n"));
        line.erase(line.find_last_not_of(" \t\r\n") + 1);
        if (!line.empty()) lines.push_back(line);
    }
    file.close();

    if (lines.size() < 3) {
        cerr << "Error: Invalid file format" << endl;
        exit(1);
    }

    istringstream iss0(lines[0]);
    iss0 >> instance.n_items >> instance.capacity;

    istringstream iss1(lines[1]);
    int weight;
    while (iss1 >> weight) instance.weights.push_back(weight);

    istringstream iss2(lines[2]);
    int value;
    while (iss2 >> value) instance.values.push_back(value);

    if ((int)instance.weights.size() != instance.n_items ||
        (int)instance.values.size()  != instance.n_items) {
        cerr << "Error: weight/value count mismatch with n_items" << endl;
        exit(1);
    }

    return instance;
}

namespace {
    // Search state, scoped to a single solve.
    int N;
    int CAP;
    const int* W;     // sorted weights
    const int* V;     // sorted values
    int best_value;
    vector<char> cur_mask;
    vector<char> best_mask;

    double upper_bound(int level, int cur_weight, int cur_value) {
        double bound = cur_value;
        int remaining = CAP - cur_weight;
        for (int i = level; i < N; ++i) {
            if (W[i] <= remaining) {
                bound += V[i];
                remaining -= W[i];
            } else {
                bound += (double)V[i] * remaining / W[i];
                break;
            }
        }
        return bound;
    }

    void branch(int level, int cur_weight, int cur_value) {
        if (cur_value > best_value) {
            best_value = cur_value;
            best_mask = cur_mask;
        }
        if (level == N) return;
        if (upper_bound(level, cur_weight, cur_value) <= best_value) return;

        if (cur_weight + W[level] <= CAP) {
            cur_mask[level] = 1;
            branch(level + 1, cur_weight + W[level], cur_value + V[level]);
            cur_mask[level] = 0;
        }
        branch(level + 1, cur_weight, cur_value);
    }
}

Solution knapsack_branch_bound(const vector<int>& weights,
                               const vector<int>& values,
                               int capacity) {
    int n = (int)weights.size();
    Solution sol{0, 0, {}, 0.0};

    if (n == 0 || capacity <= 0) return sol;

    // Sort indices by value/weight ratio descending.
    vector<int> order(n);
    iota(order.begin(), order.end(), 0);
    sort(order.begin(), order.end(), [&](int a, int b) {
        return (double)values[a] / weights[a] > (double)values[b] / weights[b];
    });

    vector<int> sw(n), sv(n);
    for (int i = 0; i < n; ++i) {
        sw[i] = weights[order[i]];
        sv[i] = values[order[i]];
    }

    N = n;
    CAP = capacity;
    W = sw.data();
    V = sv.data();
    best_value = 0;
    cur_mask.assign(n, 0);
    best_mask.assign(n, 0);

    branch(0, 0, 0);

    sol.max_value = best_value;
    for (int i = 0; i < n; ++i) {
        if (best_mask[i]) {
            sol.selected_items.push_back(order[i]);
            sol.total_weight += weights[order[i]];
        }
    }
    sort(sol.selected_items.begin(), sol.selected_items.end());
    return sol;
}

Solution solve_instance(const string& filepath, bool verbose = true) {
    KnapsackInstance instance = read_knapsack_instance(filepath);

    if (verbose) {
        cout << "\n" << string(60, '=') << endl;
        cout << "Solving instance: " << filepath << endl;
        cout << string(60, '=') << endl;
        cout << "Number of items: " << instance.n_items << endl;
        cout << "Capacity: " << instance.capacity << endl;
    }

    auto start_time = chrono::high_resolution_clock::now();
    Solution solution = knapsack_branch_bound(instance.weights, instance.values, instance.capacity);
    auto end_time = chrono::high_resolution_clock::now();

    chrono::duration<double> elapsed = end_time - start_time;
    solution.time_seconds = elapsed.count();

    if (verbose) {
        cout << "\n" << string(60, '-') << endl;
        cout << "SOLUTION" << endl;
        cout << string(60, '-') << endl;
        cout << "Optimal value: " << solution.max_value << endl;
        cout << "Total weight: " << solution.total_weight << "/" << instance.capacity << endl;
        cout << "Items selected: " << solution.selected_items.size() << "/" << instance.n_items << endl;

        cout << "Selected item indices: [";
        for (size_t i = 0; i < solution.selected_items.size(); ++i) {
            cout << solution.selected_items[i];
            if (i + 1 < solution.selected_items.size()) cout << ", ";
        }
        cout << "]" << endl;

        cout << "\nItem details:" << endl;
        cout << "  " << left << setw(6) << "Item"
             << setw(8) << "Weight" << setw(8) << "Value" << endl;
        cout << "  " << string(6, '-') << " " << string(8, '-')
             << " " << string(8, '-') << endl;
        for (int idx : solution.selected_items) {
            cout << "  " << left << setw(6) << idx
                 << setw(8) << instance.weights[idx]
                 << setw(8) << instance.values[idx] << endl;
        }

        cout << "\n" << string(60, '-') << endl;
        cout << "Computation time: " << fixed << setprecision(6)
             << solution.time_seconds << " seconds" << endl;
        cout << string(60, '=') << "\n" << endl;
    }

    return solution;
}

int main() {
    vector<string> instance_files = {
        "notebooks/plan_md_instance.txt",
        "notebooks/instance_n10.txt",
        "notebooks/instance_n15.txt",
        "notebooks/instance_n20.txt",
        "notebooks/instance_n22.txt",
        "notebooks/instance_n25.txt"
    };

    vector<Solution> results;
    vector<string> successful_files;

    cout << "\n" << string(60, '=') << endl;
    cout << "0-1 KNAPSACK PROBLEM - BRANCH AND BOUND SOLVER (C++)" << endl;
    cout << string(60, '=') << endl;

    for (const string& f : instance_files) {
        ifstream test_file(f);
        if (test_file.good()) {
            test_file.close();
            try {
                Solution r = solve_instance(f, true);
                results.push_back(r);
                successful_files.push_back(f);
            } catch (const exception& e) {
                cerr << "Error solving " << f << ": " << e.what() << "\n" << endl;
            }
        } else {
            cout << "File not found: " << f << "\n" << endl;
        }
    }

    if (!results.empty()) {
        cout << "\n" << string(60, '=') << endl;
        cout << "SUMMARY" << endl;
        cout << string(60, '=') << endl;
        cout << left << setw(25) << "Instance"
             << setw(4) << "n"
             << setw(8) << "Opt Val"
             << setw(12) << "Time (s)" << endl;
        cout << string(60, '-') << endl;

        for (size_t i = 0; i < results.size(); ++i) {
            string filename = successful_files[i];
            size_t last_slash = filename.find_last_of("/\\");
            if (last_slash != string::npos) filename = filename.substr(last_slash + 1);

            KnapsackInstance inst = read_knapsack_instance(successful_files[i]);
            cout << left << setw(25) << filename
                 << setw(4) << inst.n_items
                 << setw(8) << results[i].max_value
                 << setw(12) << fixed << setprecision(6) << results[i].time_seconds << endl;
        }
        cout << string(60, '=') << "\n" << endl;
    }

    return 0;
}

File: src/knapsack_branch_bound.py
"""
Branch and Bound Solver for 0-1 Knapsack Problem

Replaces the O(2^n) brute force enumeration with a DFS Branch and Bound
search guided by a linear (LP) relaxation upper bound.

Pipeline:
    1. Items are pre-sorted by value/weight ratio descending.
    2. DFS tries "include" before "exclude" so a strong incumbent appears early.
    3. Bound = current_value + fractional knapsack over remaining items.
       If bound <= best_value, the subtree is pruned.
    4. Infeasible (weight > capacity) branches are pruned immediately.

Worst case is still O(2^n), but the bound typically collapses the search
by orders of magnitude on non-pathological instances.
"""

from typing import List, Tuple
import time


def read_knapsack_instance(filepath: str) -> Tuple[int, int, List[int], List[int]]:
    """
    Read a knapsack instance from a text file. Accepts either format:

        3-line:
            Line 1: n_items capacity
            Line 2: weight_1 ... weight_n
            Line 3: value_1 ... value_n

        4-line (canonical, see data/knapsack_input.txt):
            Line 1: n_items
            Line 2: weight_1 ... weight_n
            Line 3: value_1 ... value_n
            Line 4: capacity
    """
    with open(filepath, 'r') as f:
        lines = [line.strip() for line in f.readlines() if line.strip()]

    if len(lines) == 4:
        n_items = int(lines[0])
        weights = list(map(int, lines[1].split()))
        values = list(map(int, lines[2].split()))
        capacity = int(lines[3])
    elif len(lines) == 3:
        first = lines[0].split()
        n_items = int(first[0])
        capacity = int(first[1])
        weights = list(map(int, lines[1].split()))
        values = list(map(int, lines[2].split()))
    else:
        raise ValueError(
            f"Expected 3 or 4 non-empty lines in {filepath}, got {len(lines)}"
        )

    assert len(weights) == n_items, f"Expected {n_items} weights, got {len(weights)}"
    assert len(values) == n_items, f"Expected {n_items} values, got {len(values)}"

    return n_items, capacity, weights, values


def knapsack_branch_bound(weights: List[int], values: List[int], capacity: int) -> Tuple[int, List[int]]:
    """
    Solve 0-1 knapsack using DFS Branch and Bound with LP-relaxation bounding.

    Returns (max_value, selected_items) where selected_items uses the
    caller's original item indices (not the internal sorted order).
    """
    n = len(weights)
    if n == 0 or capacity <= 0:
        return 0, []

    order = sorted(range(n), key=lambda i: values[i] / weights[i], reverse=True)
    w = [weights[i] for i in order]
    v = [values[i] for i in order]

    best_value = 0
    best_mask = [False] * n
    cur_mask = [False] * n

    def upper_bound(level: int, cur_weight: int, cur_value: int) -> float:
        bound = cur_value
        remaining = capacity - cur_weight
        for i in range(level, n):
            if w[i] <= remaining:
                bound += v[i]
                remaining -= w[i]
            else:
                bound += v[i] * remaining / w[i]
                break
        return bound

    def branch(level: int, cur_weight: int, cur_value: int) -> None:
        nonlocal best_value, best_mask

        if cur_value > best_value:
            best_value = cur_value
            best_mask = cur_mask.copy()

        if level == n:
            return

        if upper_bound(level, cur_weight, cur_value) <= best_value:
            return

        if cur_weight + w[level] <= capacity:
            cur_mask[level] = True
            branch(level + 1, cur_weight + w[level], cur_value + v[level])
            cur_mask[level] = False

        branch(level + 1, cur_weight, cur_value)

    branch(0, 0, 0)

    selected_original = sorted(order[i] for i in range(n) if best_mask[i])
    return best_value, selected_original


def solve_instance(filepath: str, verbose: bool = True) -> dict:
    """
    Solve a knapsack instance from file and return detailed results.
    """
    n_items, capacity, weights, values = read_knapsack_instance(filepath)

    if verbose:
        print(f"\n{'='*60}")
        print(f"Solving instance: {filepath}")
        print(f"{'='*60}")
        print(f"Number of items: {n_items}")
        print(f"Capacity: {capacity}")

    start_time = time.perf_counter()
    max_value, selected_items = knapsack_branch_bound(weights, values, capacity)
    elapsed_time = time.perf_counter() - start_time

    total_weight = sum(weights[i] for i in selected_items)
    n_selected = len(selected_items)

    if verbose:
        print(f"\n{'─'*60}")
        print(f"SOLUTION")
        print(f"{'─'*60}")
        print(f"Optimal value: {max_value}")
        print(f"Total weight: {total_weight}/{capacity}")
        print(f"Items selected: {n_selected}/{n_items}")
        print(f"Selected item indices: {selected_items}")
        print(f"\nItem details:")
        print(f"  {'Item':<6} {'Weight':<8} {'Value':<8}")
        print(f"  {'-'*6} {'-'*8} {'-'*8}")
        for idx in selected_items:
            print(f"  {idx:<6} {weights[idx]:<8} {values[idx]:<8}")
        print(f"\n{'─'*60}")
        print(f"Computation time: {elapsed_time:.6f} seconds")
        print(f"{'='*60}\n")

    return {
        'instance': filepath,
        'n_items': n_items,
        'capacity': capacity,
        'optimal_value': max_value,
        'total_weight': total_weight,
        'n_selected': n_selected,
        'selected_items': selected_items,
        'time_seconds': elapsed_time,
        'weights': weights,
        'values': values,
    }


def main():
    import os

    instance_files = [
        'notebooks/plan_md_instance.txt',
        'notebooks/instance_n10.txt',
        'notebooks/instance_n15.txt',
        'notebooks/instance_n20.txt',
        'notebooks/instance_n22.txt',
        'notebooks/instance_n25.txt',
    ]

    results = []

    print("\n" + "=" * 60)
    print("0-1 KNAPSACK PROBLEM - BRANCH AND BOUND SOLVER (Python)")
    print("=" * 60)

    for instance_file in instance_files:
        if os.path.exists(instance_file):
            try:
                results.append(solve_instance(instance_file, verbose=True))
            except Exception as e:
                print(f"Error solving {instance_file}: {e}\n")
        else:
            print(f"File not found: {instance_file}\n")

    if results:
        print("\n" + "=" * 60)
        print("SUMMARY")
        print("=" * 60)
        print(f"{'Instance':<25} {'n':<4} {'Opt Val':<8} {'Time (s)':<12}")
        print("-" * 60)
        for r in results:
            print(f"{os.path.basename(r['instance']):<25} {r['n_items']:<4} "
                  f"{r['optimal_value']:<8} {r['time_seconds']:<12.6f}")
        print("=" * 60 + "\n")


if __name__ == "__main__":
    main()
